sdk/core/azure-core/src/test/java/com/azure/core/util/serializer/JacksonAdapterIT.java

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.core.util.serializer;
+
+import com.azure.core.http.HttpHeaders;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+/**
+ * Integration tests for {@link JacksonAdapter}.
+ */
+public class JacksonAdapterIT {
+    @Test
+    public void stronglyTypedHeadersClassIsLazilyDeserialized() throws IOException {
+        final String expectedValue = "the-string-value";
+
+        HttpHeaders rawHeaders = new HttpHeaders().set("the-string", "the-string-value");
+
+        StronglyTypedHeaders stronglyTypedHeaders = JacksonAdapter.createDefaultSerializerAdapter()
+            .deserialize(rawHeaders, StronglyTypedHeaders.class);
+
+        assertNull(stronglyTypedHeaders.theString);
+        assertEquals(expectedValue, stronglyTypedHeaders.getTheString());
+    }
+
+    private static class StronglyTypedHeaders {
+        private final HttpHeaders rawHeaders;
+
+        boolean hasTheStringBeenDeserialized;
+
+        String theString;
+
+        StronglyTypedHeaders(HttpHeaders rawHeaders) {
```

Comment:
Looking for opinions on the design changes to support deferred deserialization.

Right now, I've added a constructor that accepts `HttpHeaders` but it will need to be public for `ReflectionUtils` to properly access it in Java 8 without "cracking" scopes to access non-public members. I'm not entirely convinces on a public constructor, instead I was thinking a factory method could be used, this could allow the implicit empty constructor to continue being available, so if there is any reflection access issues that occur the old fallback of using `HttpHeaders` -> JSON tree -> strongly-typed headers class would continue working. This would require an additional check in the getter to see if `rawHeaders == null` to short-circuit on returning the value without attempting deserialization or an explicit empty constructor could be added that sets a flag that eager deserialization was used. Below is a code sample of the idea:

```java
public class StronglyTypedHeaders {
  private final HttpHeaders rawHeaders;
  private final boolean eagerDeserializationUsed;

  private boolean hasTheStringBeenDeserialized;

  @JsonProperty("the-string")
  private String theString;
  
  // Explicit empty constructor to be used by Jackson.
  private StronglyTypedHeaders() {
    this.rawHeaders = null;
    this.eagerDeserializationUsed = true;
  }

  // Public constructor to be used by azure-core
  public StronglyTypedHeaders(HttpHeaders rawHeaders) {
    this.rawHeaders = rawHeaders;
    this.eagerDeserializationUsed = false;
  }

  public String getTheString() {
    if (!hasTheStringBeenDeserialized && !eagerDeserializationUsed) {
      theString = rawHeaders.getValue("the-string");
      hasTheStringBeenDeserialized = true;
    }

    return theString;
  }
}
```

Reply:
I think public ctor is fine.